src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/TagBuilder.cs

```diff
@@ -252,6 +251,15 @@ public void MergeAttribute(string key, string value, bool replaceExisting)
         /// <inheritdoc />
         public void WriteTo(TextWriter writer, HtmlEncoder encoder)
         {
+            var htmlTextWriter = writer as HtmlTextWriter;
+            if (htmlTextWriter != null)
+            {
+                // As a perf optimization, we can buffer this output rather than writing it
+                // out character by character.
+                htmlTextWriter.Write(this);
+                return;
```

Comment:
This saves
1. a `StringWriter`
2. a `StringBuilder` that backs the `StringWriter`
3. a `char[]` that backs the `StringBuilder`
4. a `string` that gets allocated when `StringWriter.ToString()` is called.